common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	payload := &commonpb.Payload{}
+	*payloadRef = payload
+	payload.Metadata = make(map[string][]byte)
+	payload.Data = content.Data
+
+	if !isStandardNexusContent(content) {
+		for k, v := range content.Header {
+			payload.Metadata[k] = []byte(v)
+		}
+		payload.Metadata["encoding"] = []byte("unknown/nexus-content")
+		return nil
+	}
+
+	contentType := content.Header.Get("type")
+	if contentType == "" {
+		payload.Metadata["encoding"] = []byte("binary/null")
+		return nil
+	}
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return err
+	}
+
+	switch mediaType {
+	case "application/x-temporal-payload":
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "application/json":
+		if params["format"] == "protobuf" {
+			payload.Metadata["encoding"] = []byte("json/protobuf")
+			messageType := params["message-type"]
+			if messageType != "" {
+				payload.Metadata["messageType"] = []byte(messageType)
+			}
+		} else {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		}
+	case "application/x-protobuf":
+		payload.Metadata["encoding"] = []byte("binary/protobuf")
+		messageType := params["message-type"]
+		if messageType != "" {
+			payload.Metadata["messageType"] = []byte(messageType)
+		}
+	case "application/octet-stream":
+		payload.Metadata["encoding"] = []byte("binary/plain")
+	default:
+		return fmt.Errorf("%w: standard content detection failed", errSerializer)
+	}
+	return nil
+}
+
+func isStandardNexusContent(content *nexus.Content) bool {
+	h := content.Header
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		return false
+	}
+	contentType := h.Get("type")
+	if contentType == "" {
+		return len(h) == 0 && len(content.Data) == 0
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return false
+	}
+
+	switch mediaType {
+	case "application/octet-stream",
+		"application/x-temporal-payload":
+		return len(params) == 0
+	case "application/json":
+		if params["format"] == "protobuf" {
+			if params["message-type"] != "" {
+				return len(params) == 2
+			}
+			return len(params) == 1
+		}
+		return len(params) == 0
+	case "application/x-protobuf":
+		if params["message-type"] != "" {
+			return len(params) == 1
+		}
+		return len(params) == 0
+	}
+	return false
+}
+
+// Serialize implements nexus.Serializer.
+func (nexusPayloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	payload, ok := v.(*commonpb.Payload)
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot serialize %v", errSerializer, v)
+	}
+
+	if payload == nil {
+		return &nexus.Content{Header: nexus.Header{}}, nil
+	}
+
+	if !isStandardPayload(payload) {
+		data, err := payload.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("%w: payload marshal error: %w", errSerializer, err)
```

Comment:
You may as well just have `errSerializer` be a string prefix; it doesn't look like we gain anything by wrapping it into the error tree

Reply:
Mostly did this to satisfy our linter.

Reply:
Again, I could make a linter exception here. I was just "following the rules".